### ISSUE TITLE
Bug in quaternion conversion?

### ISSUE THREAD

**Issue by versatran01** [utiasSTARS/liegroups#1]
Code to reproduce
```
R = np.array([[-0.26588199,  0.0721791,  -0.96129961],
                      [ 0.01772801,  0.99739046,  0.06998566],
                      [ 0.96384256,  0.001566,   -0.26646775]])
R = SO3.from_matrix(R, True)
q = R.to_quaternion()
print(np.isclose(la.norm(q), 1))
```
prints False

**Comment by versatran01** [comment]
Possible fix
```
ef uquat_from_rot3(R, fast=False):
    if fast:
        return _uquat_from_rot3_itzhack(R)
    else:
        return _uquat_from_rot3_klumpp(R)


def _uquat_from_rot3_klumpp(R):
    """
    Unit quaternion from rotation matrix, Klumpp's method
    http://arc.aiaa.org/doi/abs/10.2514/3.27947
    http://www.euclideanspace.com/comm/geom/rotations/conversions/matrixToQuaternion/index.htm
    """
    ((r00, r01, r02), (r10, r11, r12), (r20, r21, r22)) = R

    tr = np.trace(R)

    if tr > 0:
        # s = 4 * w
        s = np.sqrt(tr + 1.0) * 2
        w = s / 4.0
        x = (r21 - r12) / s
        y = (r02 - r20) / s
        z = (r10 - r01) / s
    elif r00 > r11 and r00 > r22:
        # s = 4 * x
        s = np.sqrt(1 + r00 - r11 - r22) * 2.0
        w = (r21 - r12) / s
        x = s / 4.0
        y = (r01 + r10) / s
        z = (r02 + r20) / s
    elif r11 > r22:
        # s = 4 * y
        s = np.sqrt(1 + r11 - r00 - r22) * 2.0
        w = (r02 - r20) / s
        x = (r01 + r10) / s
        y = s / 4.0
        z = (r12 + r21) / s
    else:
        # s = 4 * z
        s = np.sqrt(1 + r22 - r00 - r11) * 2
        w = (r10 - r01) / s
        x = (r02 + r20) / s
        y = (r12 + r21) / s
        z = s / 4.0

    q = np.array([w, x, y, z], float)
    if w < 0.0:
        np.negative(q, q)

    return q


def _uquat_from_rot3_itzhack(R):
    """
    New method for extracting the quaternion from a rotation matrix
    Itzhack Y. Bar-Itzhack
    This is only the precise version
    """
    ((r00, r01, r02), (r10, r11, r12), (r20, r21, r22)) = R

    # K = 0.5 * np.array([[r00 - r11, r10 + r01, r20, -r21],
    #                     [r10 + r01, r11 - r00, r21, r20],
    #                     [r20, r21, -r00 - r11, r01 - r10],
    #                     [-r21, r20, r12 - r21, r00 + r11]])
    # NOTE: eigh vs eig
    # quaternion is eigenvector of K that corresponds to eigenvalue of 1
    # Although in itzhack's paper he mentioned that "for a precise DCM,
    # there is no need to use three pairs because the result will be
    # identical", in practice, we choose to compute K using 3 pairs,
    # because the resulting K is a symmetric matrix and according to
    # http://mail.scipy.org/pipermail/numpy-discussion/2006-March/019618.html
    # eigh() is faster, and numerically more reliable than eig()

    K = np.array([[r00 - r11 - r22, 0, 0, 0],
                  [r01 + r10, r11 - r00 - r22, 0, 0],
                  [r02 + r20, r12 + r21, r22 - r00 - r11, 0],
                  [r21 - r12, r02 - r20, r10 - r01, r00 + r11 + r22]],
                 dtype=float) / 3.0
    # We use the lower triangular part of the K matrix
    eig_vals, eig_vecs = la.eigh(K, UPLO='L')
    q = eig_vecs[:, np.argmax(eig_vals)]
    # Convert to hamilton quaternion
    q = q[[3, 0, 1, 2]]

    # When computing the eigenvector of the largest eigenvalue of K, one may
    # obtain either q or -q. Usually we want the smallest rotation about
    # this axis, so we simply choose the quaternion whose scalar component
    # is positive
    if q[0] < 0.0:
        np.negative(q, q)

    return q
```

**Comment by valentinp** [comment]
Just an FYI: this has now been fixed in a08c9a2c23da579dd1dec8e087a9c49f8fccd7bf.
